File: lib/python/secrets_sdk/exceptions.py
"""Custom exceptions for the secrets SDK."""

from __future__ import annotations


class SecretsSDKError(Exception):
    """Base exception for all SDK errors."""


class VaultError(SecretsSDKError):
    """Base exception for Vault-related errors."""


class VaultAuthError(VaultError):
    """Raised when Vault authentication fails.

    Common causes: expired token, invalid OIDC callback, bad AppRole
    credentials, or missing authentication configuration.
    """

    def __init__(self, method: str, detail: str = "") -> None:
        self.method = method
        self.detail = detail
        msg = f"Vault authentication failed using {method}"
        if detail:
            msg += f": {detail}"
        super().__init__(msg)


class VaultSecretNotFound(VaultError):
    """Raised when a requested secret path does not exist in Vault."""

    def __init__(self, path: str) -> None:
        self.path = path
        super().__init__(f"Secret not found at path: {path}")


class VaultConnectionError(VaultError):
    """Raised when Vault is unreachable."""

    def __init__(self, addr: str, detail: str = "") -> None:
        self.addr = addr
        self.detail = detail
        msg = f"Cannot connect to Vault at {addr}"
        if detail:
            msg += f": {detail}"
        super().__init__(msg)


class VaultLeaseError(VaultError):
    """Raised when a lease operation fails (renew, revoke)."""

    def __init__(self, lease_id: str, operation: str, detail: str = "") -> None:
        self.lease_id = lease_id
        self.operation = operation
        self.detail = detail
        msg = f"Lease {operation} failed for {lease_id}"
        if detail:
            msg += f": {detail}"
        super().__init__(msg)


class SopsError(SecretsSDKError):
    """Base exception for SOPS-related errors."""


class SopsDecryptError(SopsError):
    """Raised when SOPS decryption fails."""

    def __init__(self, path: str, detail: str = "") -> None:
        self.path = path
        self.detail = detail
        msg = f"SOPS decryption failed for {path}"
        if detail:
            msg += f": {detail}"
        super().__init__(msg)


class SopsEncryptError(SopsError):
    """Raised when SOPS encryption fails."""

    def __init__(self, path: str, detail: str = "") -> None:
        self.path = path
        self.detail = detail
        msg = f"SOPS encryption failed for {path}"
        if detail:
            msg += f": {detail}"
        super().__init__(msg)


class SopsNotInstalledError(SopsError):
    """Raised when the sops binary is not found on PATH."""

    def __init__(self) -> None:
        super().__init__(
            "sops binary not found on PATH. Install from https://github.com/getsops/sops"
        )


class ConfigValidationError(SecretsSDKError):
    """Raised when configuration validation fails."""

    def __init__(self, issues: list[str]) -> None:
        self.issues = issues
        count = len(issues)
        summary = f"Configuration validation found {count} issue{'s' if count != 1 else ''}"
        detail = "\n  - ".join([""] + issues)
        super().__init__(f"{summary}:{detail}")


class RotationError(SecretsSDKError):
    """Raised when a secret rotation operation fails."""

    def __init__(self, detail: str) -> None:
        self.detail = detail
        super().__init__(f"Secret rotation failed: {detail}")

File: lib/typescript/src/exceptions.ts
/**
 * Custom error classes for the secrets SDK.
 *
 * Mirrors the Python SDK exception hierarchy.
 */

export class SecretsSDKError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "SecretsSDKError";
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

export class VaultError extends SecretsSDKError {
  constructor(message: string) {
    super(message);
    this.name = "VaultError";
  }
}

export class VaultAuthError extends VaultError {
  public readonly method: string;
  public readonly detail: string;

  constructor(method: string, detail: string = "") {
    const msg = detail
      ? `Vault authentication failed using ${method}: ${detail}`
      : `Vault authentication failed using ${method}`;
    super(msg);
    this.name = "VaultAuthError";
    this.method = method;
    this.detail = detail;
  }
}

export class VaultSecretNotFound extends VaultError {
  public readonly path: string;

  constructor(path: string) {
    super(`Secret not found at path: ${path}`);
    this.name = "VaultSecretNotFound";
    this.path = path;
  }
}

export class VaultConnectionError extends VaultError {
  public readonly addr: string;
  public readonly detail: string;

  constructor(addr: string, detail: string = "") {
    const msg = detail
      ? `Cannot connect to Vault at ${addr}: ${detail}`
      : `Cannot connect to Vault at ${addr}`;
    super(msg);
    this.name = "VaultConnectionError";
    this.addr = addr;
    this.detail = detail;
  }
}

export class VaultLeaseError extends VaultError {
  public readonly leaseId: string;
  public readonly operation: string;
  public readonly detail: string;

  constructor(leaseId: string, operation: string, detail: string = "") {
    const msg = detail
      ? `Lease ${operation} failed for ${leaseId}: ${detail}`
      : `Lease ${operation} failed for ${leaseId}`;
    super(msg);
    this.name = "VaultLeaseError";
    this.leaseId = leaseId;
    this.operation = operation;
    this.detail = detail;
  }
}

export class SopsError extends SecretsSDKError {
  constructor(message: string) {
    super(message);
    this.name = "SopsError";
  }
}

export class SopsDecryptError extends SopsError {
  public readonly path: string;
  public readonly detail: string;

  constructor(path: string, detail: string = "") {
    const msg = detail
      ? `SOPS decryption failed for ${path}: ${detail}`
      : `SOPS decryption failed for ${path}`;
    super(msg);
    this.name = "SopsDecryptError";
    this.path = path;
    this.detail = detail;
  }
}

export class SopsEncryptError extends SopsError {
  public readonly path: string;
  public readonly detail: string;

  constructor(path: string, detail: string = "") {
    const msg = detail
      ? `SOPS encryption failed for ${path}: ${detail}`
      : `SOPS encryption failed for ${path}`;
    super(msg);
    this.name = "SopsEncryptError";
    this.path = path;
    this.detail = detail;
  }
}

export class SopsNotInstalledError extends SopsError {
  constructor() {
    super(
      "sops binary not found on PATH. Install from https://github.com/getsops/sops"
    );
    this.name = "SopsNotInstalledError";
  }
}

export class ConfigValidationError extends SecretsSDKError {
  public readonly issues: string[];

  constructor(issues: string[]) {
    const count = issues.length;
    const summary = `Configuration validation found ${count} issue${count !== 1 ? "s" : ""}`;
    const detail = issues.map((i) => `  - ${i}`).join("\n");
    super(`${summary}:\n${detail}`);
    this.name = "ConfigValidationError";
    this.issues = issues;
  }
}

export class RotationError extends SecretsSDKError {
  public readonly detail: string;

  constructor(detail: string) {
    super(`Secret rotation failed: ${detail}`);
    this.name = "RotationError";
    this.detail = detail;
  }
}
